### ISSUE TITLE
PHP Target

### ISSUE THREAD

**Issue by marcospassos** [antlr/antlr4#1850]
I would like to propose introducing a PHP target and runtime.

Is anyone interested in joining me to implement a PHP target?

**Comment by ericvergnaud** [comment]
Hi,

you might want to also look for contributors on the google group.

Eric

**Comment by marcospassos** [comment]
Done!
https://groups.google.com/d/msg/antlr-discussion/tTzisGfE-wI/b1745axHAwAJ

**Comment by DHager** [comment]
Wow, nostalgia -- I briefly experimented with a PHP code-generation target back around '08, but that would be for ANTLR3 and everything's probably very different in v4.

My main concern would be that "If you build it, they will come" has backfired on me in the past. Is there enough demand for a PHP runtime?

**Comment by marcospassos** [comment]
I'm not sure. However, it's a real demand that we're addressing by now using a PHP/Java bridge. It requires a communication bridge, serialization protocols, and other complications. It would be a lot easier using PHP runtime (although I'm not sure if a PHP Runtime would be faster than the approach previously mentioned).

**Comment by marcospassos** [comment]
I'm not sure.

2017-06-15 22:23 GMT-03:00 Darien <notifications@github.com>:

**Comment by DHager** [comment]
I started playing around with the code yesterday, still getting a feel for the project layout, Maven machinery, and the API for runtimes. I think the next step -- if I value my sanity --  will be a Docker image that has enough Java and PHP7 scaffolding in it to do all the builds and tests in one environment.

**Comment by ericvergnaud** [comment]
Yes. Please note that we have a Travis build, so you might want to anticipate on final integration

**Comment by marcospassos** [comment]
@DHager Cool! How about PHP 7.1+? Can you share the repository link?

**Comment by DHager** [comment]
@marcospassos Currently targeting 7.1, the messy feature branch is [here](https://github.com/DHager/antlr4/tree/runtime-php7). So far my process has been to run [some of these steps](https://github.com/DHager/antlr4/blob/runtime-php7/runtime/Php7/docker/readme.md) while tuning the `.stg` files, gradually changing the generated code (e.g. `HelloLexer.php`) into something resembling PHP7 syntax with docblocks to capture type-information.  

Right now my philosophy is that it's better to make something that closely resembles the Java scaffolding and diffs nicely, as opposed to redesigning the API to make it "safer" to counteract some of the type-checking/read-only guarantees that are being lost in the conversion. 

I'm holding off on writing "real PHP files" (in `antlr4/runtime/Php7/src`) because there are some semantic differences to hammer out that might significantly change what they look like. In the immediate future, I'm thinking about stuff like:

* Distinguishing some places where Java code isn't explicit about the difference between instance/static access, but PHP needs to be.
* Changing static properties (with non-constants) to getter-methods.
* Change generation of inner classes to instead create full classes in *other files*, that live in a PHP namespace named after the former outer-class. Ensuring the former outer/inner classes can still collaborate correctly.
* Going through static-access situations where late-static-binding is (or isn't) needed. (`self::$x` vs `static::$x`)
* Replace the use of `List<?>` with a PHP equivalent *object* in the short term (to maintain the same copy/write semantics) and then later figure out where it's safe to use a regular `array()` instead for performance.

On the Docker/environment side, putting in XDebug is a matter of [copy pasting from another project](https://github.com/DHager/docker-php-library-demo), but I have much less experience with remote Java debugging.

**Comment by DHager** [comment]
@ericvergnaud Thanks for the heads-up. I'm not very familiar with the Travis stuff, but I don't think it'll present any new quirks beyond whatever the system is already handling for the other targets: Since this is just command-line PHP, a whole universe of web-weirdness can be avoided.

**Comment by DHager** [comment]
Actually, there's one big question that I don't know enough about the long-term evolution of Antlr4 to answer: Is it better to use the Java runtime as a model, or is it better to adapt from the Python(2/3) runtime? I'm sure in both cases there are aspects that simply won't translate to PHP.

Reasons to crib from Java:

* Primary implementation, similarity will likely mean easier changes in the future
* Has interfaces, and class access-control keywords which are similar to Java's
* A limited amount of the type-checking can be kept in the form of parameter type-hints and doc-blocks to aid IDEs

Reasons to crib from Python:

* Already stripped out a lot of type-checking code
* Already avoids function overloading
* Probably already solved a bunch of lurking issues involves in avoiding Java-specific features

**Comment by ericvergnaud** [comment]
FYI I created the Python 3 runtime first to keep the type hints, then Python 2 from Python 3, then JavaScript from Python 2

Envoyé de mon iPhone

**Comment by DHager** [comment]
I think I need help with the codegen/StringTemplate workings, specifically involving how ANTLR emits the `StructDecl` templates. 

The big issue is that PHP does not support inner classes, nor does it support defining an anonymous class at the same time you assign it to a static variable. (Which would also prevent you from using the anonymous class in type-hinting and documentation.)

I'd prefer to move the `ParserRuleContext` subclasses into their own files in a subdirectory (to satisfy PSR-4), but my second-choice would be to at least move them outside the main class within the same file.

### Code examples

#### Current Java structure

```java
package Example;

public class HelloParser extends Parser {
	// [...]

	public static class MyruleContext extends ParserRuleContext {
		// [...]

		@Override
		public void enterRule(ParseTreeListener listener) {
			if (listener instanceof HelloListener) ((HelloListener) listener).enterMyrule(this);
		}

		@Override
		public void exitRule(ParseTreeListener listener) {
			if (listener instanceof HelloListener) ((HelloListener) listener).exitMyrule(this);
		}
	}
	// [...]
}
```

#### Similar structure in PHP that does not work:
    
```php
namespace Example;

class HelloParser extends Parser{
    // [...]
    
    class MyruleContext extends ParserRuleContext{
        // [...]
    
        public function enterRule(ParseTreeListener $listener){
            if ($listener instanceof HelloListener) {
                $listener->enterMyrule($this);
            }
        }
    
        public function exitRule(ParseTreeListener $listener){
            if ($listener instanceof HelloListener) {
                $listener->exitMyrule($this);
            }
        }
    }
    // [...]
}
```

#### Compromise one-file PHP solution

   
```php
namespace Example;

class HelloParser extends Parser{
    // [...]
}

namespace Example\HelloParser;

class MyruleContext extends ParserRuleContext{
    // [...]
    
    public function enterRule(ParseTreeListener $listener){
        if ($listener instanceof HelloListener) {
            $listener->enterMyrule($this);
        }
    }
    
    public function exitRule(ParseTreeListener $listener){
        if ($listener instanceof HelloListener) {
            $listener->exitMyrule($this);
        }
    }
}
```

#### Ideal PHP solution with multiple files

```php
// In HelloParser.php
namespace Example;

class HelloParser extends Parser{
    // [...]
}
```

```php
// In HelloParser/MyruleContext.php
namespace Example\HelloParser;

class MyruleContext extends ParserRuleContext{
    // [...]

    public function enterRule(ParseTreeListener $listener){
        if ($listener instanceof HelloListener) {
            $listener->enterMyrule($this);
        }
    }

    public function exitRule(ParseTreeListener $listener){
        if ($listener instanceof HelloListener) {
            $listener->exitMyrule($this);
        }
    }
}
```

____

**TLDR:** I'd like to move the `StructDecl` and `AltLabelStructDecl` content into their own classes in a subdirectory, but it looks like that would require some significant changes to how the code-gen pipeline works.

**Comment by DHager** [comment]
Experimenting with compromise number 3, the anonymous classes... It feels dirty, but it does seem to avoid rewriting a bunch of the codegen/StringTemplate magic. 

In the future it might be possible to avoid the performance hit of one-class-per-instance by storing the first instance created per rule, and then spawning clones off of it, but I haven't gotten that far.

**Comment by marcospassos** [comment]
@DHager why not option 1? It seems the best option. Passing anonymous functions to construct makes the code less testable and much harder to debug.

**Comment by marcospassos** [comment]
...and disallow inheritance.

**Comment by DHager** [comment]
As I said, the first solution is my preference also, but I don't  understand enough about how the meta-magic generator code works. 

Either it's a significant rewrite of the all-languages pipeline, or else there's some existing magic that I haven't found yet.

**Comment by marcospassos** [comment]
@parrt could please give us some help here?

**Comment by ericvergnaud** [comment]
Hi,

I believe solution 1 would require a minor evolution of the antlr4 tool, similar to the one the cpp version added for header (public boolean needsHeader() { return true; }).
However you should keep in mind that this may end up in literally hundreds of files (some of which would become obsolete when rule names change)

I like solution 2 (although not being a PHP coder I don't understand the need for a per rule namespace...)
I'm not sure complying with PSR-4 is useful for contexts because contexts can only be created by the parser anyway (hence the use of inner classes in Java), and the only 2 other consumers would be listener and visitor, which being generated would know where to find these classes.

I would discourage you to implement solution 3. Being able to rapidly identify in a debugger which grammar rule relates to a rule context is crucial, and cannot be achieved with anonymous classes.  

Eric

**Comment by parrt** [comment]
In case it's relevant, @bhamiltoncx got pretty far: "*I finished about 80% of a Hack (HHVM) port of ANTLR4. Unfortunately, I didn't get to complete it, but here's the work I did:*"

https://github.com/bhamiltoncx/antlr4/tree/hack

**Comment by DHager** [comment]
@ericvergnaud For me the difference between v1 (one class per file) and v2 (concatenated class-defs) is just "nice to have". It's unlikely that "outside code" would dive that deeply into the rules, and even if they did encounter autoloading problems the workaround they'd need would be straightforward.

One reason I was pondering v1 is that it sounds like something which could be implemented as a separate PR and demoed against the Java target... assuming none of the current Java inner classes *rely* on their privileged status.

Anywho, I'll poke at v2, and see if I can at least get the Java target to ritually disembowel itself and put its structs all alone outside the Parser/Lexer class definition. If so, I'll open a separate PR to cover something to expose that behavior like `Target.needsStructsOutsideRecognizer()`.

@parrt Thanks, I'll take a look. Right now I'm focusing on the code-gen side making syntactically-valid PHP7, because I figure that's going to inform how much of the real heavy-lifting code needs to change.

**Comment by ericvergnaud** [comment]
Java inner classes do rely on their privileged status, so really not much point exploring that option for Java

**Comment by ericvergnaud** [comment]
also the JavaScript target uses standalone constructs so you might want to refer to its template to understand how this is done.

**Comment by ericvergnaud** [comment]
@parrt so following Pareto's law, @bhamiltoncx has only done 20% ;-)

**Comment by DHager** [comment]
@ericvergnaud Thanks for the tip on the JavaScript target, however it looks like it's still nesting/emitting the StringTemplate stuff in the same structure. 

In other words, it's taking advantage of how JS allows you to append new properties/methods to an prototype at any time, as opposed to languages like Java or PHP which require the class definition to be a contiguous block of code.

**Comment by ericvergnaud** [comment]
Mmm... maybe you didn't notice that rule contexts were actually defined outside the parser function ?
See for example https://github.com/prompto/prompto-javascript/blob/master/JavaScript-Core/src/generated/prompto/utils/ArgsParser.js

**Comment by DHager** [comment]
Hrm, OK, so it is not a separate template invocation, but instead printing the `funcs` parameter passed into the `Parser` template...  I can relocate the use of `that, but it moves both the inner class definition *and* the function that uses it. The real class-def occurs in the `RuleFunction` template by printing the `ruleCtx` parameter...

So maybe I need to find a way to call `RuleFunction` twice with a boolean flag, where one usage prints the class definition and the other usage prints everything else.

**Comment by DHager** [comment]
The "PS" note on my last comment seems to work for simple grammars, but throw in something more complicated and I get errors when using `<funcs.ruleCtx>` like:

    warning(22):  template error: context [/ParserFile /Parser /Parser_] 52:5 no such property or can't access: org.stringtemplate.v4.ST$AttributeList.ruleCtx

I assume it's because there's some kind of handling going on with how `org.antlr.v4.codegen.OutputModelWalker` proceses `org.antlr.v4.codegen.model.RuleFunction#ruleCtx`. Unfortunately I don't see any way to pass a true/false flag through in such a way that I can invoke `RuleFunction` twice in two different ways and places. 

So far the most-promising (albeit messy) route seems like doubling down on what the CPP target did, with a bunch of templates that have a Header-like suffix.

**Comment by DHager** [comment]
Well, I'm stuck and could use some StringTemplate guruship. Is there any way for a parent template to transmit a variable to an indirect sub-template that gets used later for rendering a property?

In a nutshell, I'm trying to change from this:

    Parser_(parser, funcs, atn, sempredFuncs, ctor, superClass) ::= <<      
        
        // Start Parser class def
        
        <funcs; separator="\n">
        
        // End Parser class def                
    >>

    RuleFunction(currentRule,args,code,locals,ruleCtx,altLabelCtxs,namedActions,finallyAction,postamble,exceptions) ::= <<
             
        <ruleCtx>               
        // Other stuff
    >>

To something like:

    Parser_(parser, funcs, atn, sempredFuncs, ctor, superClass) ::= <<      
        
        // Start Parser class def    
            
        <funcs; separator="\n"> <! TODO pass along newTemplateParameter=false somehow !>

        // End Parser class def
        
        <funcs; separator="\n"> <! TODO pass along newTemplateParameter=true somehow !>        
    >>

    RuleFunction(currentRule,args,code,locals,ruleCtx,altLabelCtxs,namedActions,finallyAction,postamble,exceptions,newTemplateParameter) ::= <<
             
        <if(newTemplateParameter)>
            <ruleCtx>
        </endif>
        
        // Other stuff
    >>

The problem is that I can't just invoke one template from another or else `OutputModelWalker` doesn't get involved which seems essential for other parameters to be populated (like `ruleCtx` itself). Conversely, there doesn't seem to be a mechanism to pass options into `<funcs>` that will somehow survive into the walker be different between both times `RuleFunction` is rendered.

**Comment by ericvergnaud** [comment]
For the prototypes, have you tried iterating over parser.funcs within the parser st?

**Comment by DHager** [comment]
@ericvergnaud Ah! Thanks, that fixed it right up -- no more template error and multiple class-defs can appear after/outside the main parser now.

I had assumed I was already tapping into some StringTemplate convenience feature that "knew" how to handle `Lists` and render each item, because it worked with the single-item `List<RuleContext>` generated by my first trivial grammar. Experimenting with some other grammars, it looks like whatever-it-was simply falls over simply due to having more rules.

    <! 
        This worked only in grammars with one rule!>
        <funcs.ruleCtx; separator="\n">
    !>

    <! Works in grammars with many rules !>
    <funcs :{ func | <func.ruleCtx> }>

**Comment by DHager** [comment]
@ericvergnaud I'm trying to get a feel for the "weight" of the supporting runtime code... Do you think it's worth building some kind of rudimentary Java-to-almost-Php7 conversion scripts, or is it better to try to adapt over all the `org.antlr.v4.runtime.*` stuff by hand?

**Comment by ericvergnaud** [comment]
I did all by hand using a tdd strategy I.e. relying on the test suite, starting with simple lexer test and gradually expanding to ultimate parser ones.
I felt that manually transcribing the code was a great opportunity to become intimate with it.
But mileage may vary.

**Comment by rstoetter** [comment]
Hello @ all

the issue from May to create a new PHP 7 target for ANTLR 4 seems still to exist. 

@DHager did you manage it to complete the target?
@marcospassos  did you start with the target in the meanwhile?

I myself started to work on a PHP 7 target, too and ran into the same problems like Darien Hager in the STG-file: PHP does not allow nested classes and there is the need for a workaround. 

Maybe the best would be, when ANTLR would not only provide the parser.funcs like in 

    Parser_(parser, funcs, atn, sempredFuncs, ctor, superClass) ::= <<      
    
        // Start Parser class def
    
        <funcs; separator="\n">
    
        // End Parser class def                
    >>

but also RuleFunction and ruleCtx as separate templates.

ANTLR provides with parser.funcs a class in ruleCtx and a function RuleFunction. There should be a possibility to use RuleFunction and ruleCtx standalone as additional templates: When doing the change one could use the parser.funcs part and extract the necessary information for RuleFunction and ruleCtx out of the calculated parser.funcs.

I suppose this would be a simple hack for somebody experienced with the ANTLR source code but I do not understand the source code enough to carry that into execution.

What I found out is, that the linchpin could be the code in OutputModelController.java 

    public void buildRuleFunction(Parser parser, Rule r) {
		RuleFunction function = rule(r);
		parser.funcs.add(function);
		pushCurrentRule(function);
		function.fillNamedActions(delegate, r);
                .
                .
                .
   }

This method seems to create the RuleFunction and the ruleCtx.


Did somebody make it to get the target working or could somebody help to do this hack?

**Comment by DHager** [comment]
@rstoetter No, unfortunately. You can see as far as I got in [https://github.com/DHager/antlr4/tree/runtime-php7](my runtime-php7 branch). I believe I managed to work around the "inner class" issue, but there's still a lot more work to do.

**Comment by PhoenixFnX** [comment]
I'm at the moment thinking about updating my antlr-3.1.3-php.jar implementation. My project if using this PHP generated parser which is working like a charm !
We are planning to move on PHP 7.2, I didn't test the generated code yet...

I'm actually SURPRISED that antlr 4 is not supporting PHP Terget anymore. That really annoys me and I'd like your opinion on how should I move onto keep using maintained libraries.

Is there a new leader on PHP Parser generators ? Using a dot.G grammar ?

Nicolas.

**Comment by marcospassos** [comment]
@DHager I pretend to start working actively in the PHP target, are you still working on it?

**Comment by DHager** [comment]
@marcospassos Nope, sorry. No particular plans to return to it, given that my new employer isn't using PHP :p

**Comment by marcospassos** [comment]
@DHager I'll continue the great work that you started if you agree.

Could you please explain how did you manage to work around the "inner class" issue?

**Comment by DHager** [comment]
@marcospassos I'm afraid I don't really remember, I suspect it had something to do with changing the templating so that the "inner classes" manifested as helper classes defined within the same file (allowed in PHP, if undesirable for autoloading).

Looking at the [changes in aggregate](https://github.com/DHager/antlr4/compare/master...DHager:runtime-php7), I suspect the relevant parts are related to `StructDecl` and `AltLabelStructDecl`, and that [this commit](https://github.com/DHager/antlr4/commit/6068ab380644427edceb9bef2d8f6455d66b8402) somehow moved them from being nested to being sequential.

**Comment by yar3333** [comment]
See https://github.com/yar3333/antlr4-php-workspace . Now only basically tested, need help to find issues. Maybe someone want to prepare and support composer version of runtime?

**Comment by marcospassos** [comment]
@yar3333 we can help improving test coverage and a composer packager.

@parrt, composer packages requires a `composer.json` file in the root of the repository. The runtime libraries, however, are under the `antlr/runtime` folder. It would require mirroring that folder to another read-only repository using `subtree split` or something or something like that. Are you ok with this approach?

**Comment by ericvergnaud** [comment]
Hi
Can’t speak for Ter but I will definitely support any TDD approach.
Éric 

Envoyé de mon iPhone

**Comment by parrt** [comment]
@marcospassos well i'm not familiar with `subtree split` but as long as the main antlr repo structure is same, you are free to fork / split or whatever. if i look in the antlr repo though, it should have all content I think.

**Comment by marcospassos** [comment]
Yes, the point is how to keep the ANTLR repo and the `php-runtime` repo in sync.

A possible workaround:
- Crate a new repository to host the PHP runtime, e.g `antlr/antlr4-php-runtime`
- Add `antlr/antlr4-php-runtime` to `antlr/antlr4/runtime/PHP` as a GIT submodule (we will embed the `antlr/antlr4-php-runtime` repository into the `antlr/antlr4/runtime/PHP` folder)
- Changes are made direct to `antlr/antlr4-php-runtime`
- Whenever the `antlr/antlr4-php-runtime` gets updated, we manually pull the changes running `git submodule update` on `antlr/antlr4`

More details can be found [here](https://gist.github.com/gitaarik/8735255).

@parrt does it work for you? If so, could please create the new `antlr/antlr4-php-runtime`?

**Comment by ericvergnaud** [comment]
Hi,
in the end state it would have to be antlr/antlr4/runtime/PHP such that any change to the tool or the test set is automatically checked against the target.
This model makes sense because what gets release is a new version of antlr4 and all supported targets (as opposed to antlr4 + java target, and other target have their own unsynched lifecycle)
this works for other targets so not sure why PHP would require anything different?

**Comment by marcospassos** [comment]
Composer, PHP's ubiquitous package manager, requires a repository per package. It means that if we place the PHP runtime at `antlr/antlr4/runtime/PHP`, all projects that depends on it will download the entire ANTLR4 codebase.

**Comment by parrt** [comment]
Hmm... not sure what to do here but I hesitate to mess up the model we've been using successfully as Eric points out.  It sounds like it's just the package manager that has the problem so maybe as part of the build process for the release, the build process could copy the stuff into a separate directory and run the tests?

**Comment by marcospassos** [comment]
It's possible using `git subtree`, but the build process should take care of keeping a read-only repository up to date, probably using a tool like [splitsh](https://github.com/splitsh/lite).

Does it sound reasonable?

**Comment by ericvergnaud** [comment]
mmmm...
I'm not a great fan of black boxes
why not add, as part of the travis build, a bash script which will on every build update this standalone repo using git commands, and then run composer against it?

**Comment by marcospassos** [comment]
> mmmm...
> I'm not a great fan of black boxes
> why not add, as part of the travis build, a bash script which will on every build update this standalone repo using git commands, and then run composer against it?

That's my suggestion (there is no need for running composer commands however, just update the standalone repository).

**Comment by ericvergnaud** [comment]
composer would have to be run when publishing a new version of antrl4, no?
does composer even need a repo?
would the below not suffice ?
 cd whatever
 git init
 git add
 git commit

**Comment by parrt** [comment]
Yep, we need a braindead set of instructions for me when pushing our a release that doesn't require weird system configuration etc...

**Comment by marcospassos** [comment]
Ok, let's move forward then!

**Comment by parrt** [comment]
@marcospassos just a reminder that this is not really my focus anymore so anything that requires thought / time from me is an issue. ;)

**Comment by yar3333** [comment]
@marcospassos I think your idea about autoupdated git repo is most simple. Composer's https://packagist.org/ have built-in function to autopublish new version of package from git repo (by URL to github) on git tag adding.

**Comment by yar3333** [comment]
So, we need 1) create new github repo for php runtime (in `antlr` account?) and 2) tune updating of this repo in travis. @marcospassos, do you have access rights to do that?

**Comment by parrt** [comment]
can we keep php target in antlr repo but do testing externally via travis? that way you don't need me for anything but PRs for php target.

**Comment by yar3333** [comment]
@parrt May be we can avoid using separate repo for php runtimes - need test (looks like `packagist` support links to subfolders).

**Comment by marcospassos** [comment]
There is no support as far as I know:
https://github.com/composer/packagist/issues/472

**Comment by yar3333** [comment]
@marcospassos You are right :(

**Comment by marcospassos** [comment]
Okay, I think I came up with a simple idea of how to move forward. We can place the runtime library at `antlr4/runtime/php` and create a fork from the ANTLR project. We then clean up the fork by keeping only the `antlr/runtime/php` folder. We use [Github's hooks](https://github.com/apps/pull) to keep the repository up to date with the ANTLR main repository.

These are the next steps I see:
- Create a fork from the antlr4 repository with the name `antlr-php-runtime`
- Review the runtime code and include some tests
- Open a PR against the ANTLR4 main repository
- Once accepted, enable automatic repository sync

@parrt the only steps you're involved are creating the repository and PR's :). Could you please create a fork from the project with name `antlr-php-runtime` and grant me access to it?

**Comment by parrt** [comment]
Fork? It won't let me fork antlr4 into antlr org as it exists already. don't see option to rename fork. should i just create antlr-php-runtime repo in antlr org and let you have your way with it?

**Comment by marcospassos** [comment]
> should i just create antlr-php-runtime repo in antlr org and let you have your way with it?
yes :)

**Comment by parrt** [comment]
and away we go! https://github.com/antlr/antlr-php-runtime

**Comment by yar3333** [comment]
I read about github webhooks now. I suggest next solution of the php package publication:

**Comment by yar3333** [comment]
I know, not ideal and I don't know Travis CI very good, but I believe what Travis CI does not allow to push into another repos, so just fixing `antlr4`'s travis config is not solution.

**Comment by marcospassos** [comment]
@parrt The PHP Target is ready. However, I could not find how to generate the tests from the stg file. Can you help me?

**Comment by parrt** [comment]
Wow! Amazing amount of work.  Well, the java code should use the templates to generate the tests as you invoke the tests (via java), if I recall.

**Comment by marcospassos** [comment]
Yeah, Months of work ^^". 

How I'm supposed to generate those files for the PHP runtime?  E.g: https://github.com/antlr/antlr4/tree/master/runtime-testsuite/test/org/antlr/v4/test/runtime/python3

Can you help me with the commands?

**Comment by parrt** [comment]
Oh, have you seen this ? https://github.com/antlr/antlr4/blob/master/doc/creating-a-language-target.md

Create a template file for runtime tests. All you have to do is provide a few templates that indicate how to print values and declare variables. Our runtime test mechanism in dir runtime-testsuite will automatically generate code using these templates for each target and check the test results. It needs to know how to define various class fields, compare members and so on. You must create a X.test.stg file underneath runtime-testsuite/resources/org/antlr/v4/test/runtime. Again, your best bet is to copy the templates from the closest language to your target and tweak it to suit.

and i think maybe those java files you mention are copied and tweaked to say your target stuff instead of 

```
BaseRuntimeTest.getRuntimeTestDescriptors(LexerErrorsDescriptors.class, "CSharp");
```

**Comment by marcospassos** [comment]
Yeah, I've seen. So, those files under https://github.com/antlr/antlr4/tree/master/runtime-testsuite/test/org/antlr/v4/test/runtime are not automatically generated? Should I copy all the java files from the closest language and adapt to PHP?

Besides this, it looks like we've to add the PHP runtime to the CI server so that it will able to run it, right?

**Comment by parrt** [comment]
Yep, copy those "connector" files and tweak to refer to your runtime. Yep, gotta modify the CI stuff so it installs anything fired up by your tests.  E.g., go runtime https://github.com/antlr/antlr4/blob/master/runtime-testsuite/test/org/antlr/v4/test/runtime/go/BaseGoTest.java has stuff like

```
String goExecutable = locateGo();
ProcessBuilder pb = new ProcessBuilder(goExecutable, "install", "-x");
```

that will refer to system executables.
